### PR TITLE
Move convergence criteria out of linear solver

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This PR moves the convergence criteria out of the linear solver so they may be used also by the nonlinear solver. In the future I can also make them generic to support other iterative algorithms, e.g. FastFlow.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsvu** [comment]
Note to reviewers: This literally just moves `LinearSolver/Convergence.{hpp,cpp}` to `Convergence/{Reason,Criteria,HasConverged}.{hpp,cpp}`.

**Comment by nilsvu** [comment]
@nilsdeppe @kidder could you review or find reviewers for this?

File: src/NumericalAlgorithms/Convergence/HasConverged.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace Convergence {
+
+/*!
+ * \brief Determine whether the \p criteria are met.
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Convergence/HasConverged.hpp:23]: It's not clear to me this function should be public vs having people use the `HasConverged` class. @nilsleiffischer do you have any reasons to prefer it being public?

File: src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitor.hpp [hunk 2]
```diff
@@ -96,13 +96,14 @@ struct InitializeResidualMonitor {
...
-      const ParallelComponent* const /*meta*/, ::Verbosity verbosity,
-      LinearSolver::ConvergenceCriteria convergence_criteria) noexcept {
+  static auto apply(const db::DataBox<tmpl::list<>>& /*box*/,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/,
+                    ::Verbosity verbosity,
+                    Convergence::Criteria convergence_criteria) noexcept {
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitor.hpp]: `const Convergence::`?

File: src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitor.hpp [hunk 3]
```diff
@@ -96,13 +96,14 @@ struct InitializeResidualMonitor {
...
-      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
-      const ParallelComponent* const /*meta*/, ::Verbosity verbosity,
-      LinearSolver::ConvergenceCriteria convergence_criteria) noexcept {
+  static auto apply(const db::DataBox<tmpl::list<>>& /*box*/,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/,
+                    ::Verbosity verbosity,
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitor.hpp]: `const ::Verbosity`?

**Comment by kidder** [comment]
@moxcodes can you review this? it should be super quick (i.e. a rename of reviewed code)

File: tests/Unit/NumericalAlgorithms/Convergence/Test_Criteria.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#include "tests/Unit/TestHelpers.hpp"
+
+SPECTRE_TEST_CASE("Unit.Numerical.Convergence.Criteria",
+                  "[Unit][NumericalAlgorithms]") {
+  const Convergence::Criteria criteria{2, 0., 0.5};
+  CHECK(criteria == Convergence::Criteria{2, 0., 0.5});
+  CHECK(criteria != Convergence::Criteria{3, 0., 0.5});
+  CHECK(criteria != Convergence::Criteria{2, 1., 0.5});
+  CHECK(criteria != Convergence::Criteria{2, 0., 0.6});
+  test_serialization(criteria);
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/Convergence/Test_Criteria.cpp:17]: Could you set the relative residual to not 0. when testing serialization, etc.?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash the changes

**Comment by moxcodes** [comment]
Sure, I can review this.

File: src/NumericalAlgorithms/Convergence/HasConverged.hpp (hunk 1 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/Convergence/HasConverged.hpp:23]: discussed this offline and decided to keep as-is

**Comment by nilsvu** [comment]
Rebased on #1408 and addressed the comments by @nilsdeppe

**Review by nilsdeppe — DISMISSED** [review]

File: src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp [hunk 5]
```diff
@@ -171,7 +173,9 @@ struct UpdateResidual {
     if (UNLIKELY(has_converged and
                  static_cast<int>(get<::Tags::Verbosity>(box)) >=
                      static_cast<int>(::Verbosity::Quiet))) {
-      Parallel::printf("%s", has_converged);
+      Parallel::printf("The linear solver has converged in %d iterations: %s",
```
Comment (moxcodes) [inline src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp]: should this have format string `%llu` or similar for the size_t ?

File: src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp [hunk 6]
```diff
@@ -236,7 +238,9 @@ struct StoreFinalOrthogonalization {
     if (UNLIKELY(has_converged and
                  static_cast<int>(get<::Tags::Verbosity>(box)) >=
                      static_cast<int>(::Verbosity::Quiet))) {
-      Parallel::printf("%s", has_converged);
+      Parallel::printf("The linear solver has converged in %d iterations: %s",
```
Comment (moxcodes) [inline src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp]: same question about the format string

**Comment by kidder** [comment]
@nilsleiffischer this isn't compiling...

File: src/NumericalAlgorithms/Convergence/HasConverged.hpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * \note This function assumes the \p iteration_id is that of the next, but
+ * not yet performed step. For instance, a `MaxIteration` criterium of 1 will
```
Comment (moxcodes) [inline src/NumericalAlgorithms/Convergence/HasConverged.hpp]: as a nitpick, I think `criterion` is preferred.

**Review by moxcodes — COMMENTED** [review]
I'm happy for you to squash any of these minor suggestions you choose to implement.

File: src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp (hunk 5 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp]: Changed to `%zu` now. I don't know exactly what the best-practice here is, so if you do please tell me.

**Review by moxcodes — DISMISSED** [review]

File: src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp (hunk 5 above)
Comment (kidder) [reply src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp]: yes, `%zu` is what I use

**Comment by nilsvu** [comment]
Rebased and fixed iwyu. This should be ready @moxcodes @nilsdeppe @sxs-collaboration/spectre-core-devs

**Review by nilsdeppe — APPROVED** [review]

**Review by moxcodes — APPROVED** [review]

**Pull request merged**